File: java/chainofresponsibility/AbstractMiddlewareHandler.java
package chainofresponsibility;

import ultimate.HttpRequest;
import ultimate.MiddlewareResponse;

public class AbstractMiddlewareHandler implements MiddlewareHandler {
    private MiddlewareHandler nextHandler;
    @Override
    public MiddlewareHandler setNext(MiddlewareHandler middlewareHandler) {
        nextHandler = middlewareHandler;
        return  middlewareHandler;
    }

    @Override
    public MiddlewareResponse handle(HttpRequest request) {
        if (nextHandler != null) {
            return nextHandler.handle(request);
        }
        return new MiddlewareResponse("Request has passed all checks successfully", true);
    }
}

File: typescript/chainofresponsibility/abstractMiddleware-handler.ts
import {MiddlewareHandler} from "./middleware-handler";
import {ChainRequest} from "./request";

export abstract class AbstractMiddlewareHandler implements MiddlewareHandler {
  private nextHandler: MiddlewareHandler;
  constructor() {
    this.nextHandler = {
      setNext: () => this,
      handle: () => true,
    };
  }

  public setNext(handler: MiddlewareHandler): MiddlewareHandler {
    this.nextHandler = handler;
    return handler;
  }

  public handle(request: ChainRequest): boolean {
    if (this.nextHandler) {
      return this.nextHandler.handle(request);
    }
    return true;
  }
}
